pkg/usagetracker/usagetrackerclient/client.go

```diff
@@ -0,0 +1,232 @@
+
+	// Allow to inject custom client factory in tests.
+	clientFactory client.PoolFactory `yaml:"-"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.StringVar(&cfg.PreferAvailabilityZone, prefix+".prefer-availability-zone", "", "Preferred availability zone to query usage-trackers.")
+	f.DurationVar(&cfg.RequestsHedgingDelay, prefix+".requests-hedging-delay", 100*time.Millisecond, "Delay before initiating requests to further usage-trackers (e.g. in other zones).")
+	f.IntVar(&cfg.ReusableWorkers, prefix+".reusable-workers", 500, "Number of pre-allocated workers used to send requests to usage-trackers. If 0, no workers pool will be used and a new goroutine will be spawned for each request.")
+
+	f.BoolVar(&cfg.TLSEnabled, prefix+".tls-enabled", cfg.TLSEnabled, "Enable TLS for gRPC client connecting to usage-tracker.")
+	cfg.TLS.RegisterFlagsWithPrefix(prefix, f)
+}
+
+type UsageTrackerClient struct {
+	services.Service
+
+	cfg    Config
+	logger log.Logger
+
+	partitionRing      *ring.PartitionInstanceRing
+	partitionBatchRing *ring.ActivePartitionBatchRing
+
+	clientsPool *client.Pool
+
+	// trackSeriesWorkersPool is the pool of workers used to send requests to usage-tracker instances.
+	trackSeriesWorkersPool *concurrency.ReusableGoroutinesPool
+}
+
+func NewUsageTrackerClient(clientName string, clientCfg Config, partitionRing *ring.PartitionInstanceRing, instanceRing *ring.Ring, logger log.Logger, registerer prometheus.Registerer) *UsageTrackerClient {
+	c := &UsageTrackerClient{
+		cfg:                    clientCfg,
+		logger:                 logger,
+		partitionRing:          partitionRing,
+		partitionBatchRing:     ring.NewActivePartitionBatchRing(partitionRing.PartitionRing()),
+		clientsPool:            newUsageTrackerClientPool(client.NewRingServiceDiscovery(instanceRing), clientName, clientCfg, logger, registerer),
+		trackSeriesWorkersPool: concurrency.NewReusableGoroutinesPool(clientCfg.ReusableWorkers),
+	}
+
+	c.Service = services.NewIdleService(nil, c.stop)
+	return c
+}
+
+// stop implements services.StoppingFn.
+func (c *UsageTrackerClient) stop(_ error) error {
+	c.trackSeriesWorkersPool.Close()
+
+	return nil
+}
+
+func (c *UsageTrackerClient) TrackSeries(ctx context.Context, userID string, series []uint64) ([]uint64, error) {
+	var (
+		batchOptions = ring.DoBatchOptions{
+			Cleanup:       nil,
+			IsClientError: func(error) bool { return false },
+			Go:            c.trackSeriesWorkersPool.Go,
+		}
+
+		rejectedMx sync.Mutex
+		rejected   []uint64
+	)
+
+	// Series hashes are 64bit but the hash ring tokens are 32bit, so we truncate
+	// hashes to 32bit to get keys to lookup in the ring.
+	keys := make([]uint32, len(series))
+	for i, hash := range series {
+		keys[i] = uint32(hash)
+	}
+
+	err := ring.DoBatchWithOptions(ctx, ring.WriteNoExtend, c.partitionBatchRing, keys,
```

Comment:
Should this be this?
```suggestion
	err := ring.DoBatchWithOptions(ctx, trackSeriesOp, c.partitionBatchRing, keys,
```
(It's the same value, but not sure if this is semantically more correct)

Reply:
>Should this be this?

Yes, definitely. It was a mistake using `ring.WriteNoExtend` here.